src/main/groovy/com/athaydes/spockframework/report/internal/HtmlReportCreator.groovy

```diff
@@ -246,37 +257,85 @@ class HtmlReportCreator extends AbstractHtmlCreator<SpecData>
 
 
     private void writeFeatureBlocks( MarkupBuilder builder, FeatureInfo feature, IterationInfo iteration = null ) {
-        for ( BlockInfo block in feature.blocks ) {
-            writeBlock( builder, block, feature, iteration )
+        if ( showCodeBlocks ){
+            writeBlocksFromCode( builder, feature, iteration )
+        } else {
+            writeBlocks( builder, feature, iteration )
         }
     }
 
-    private void writeBlock( MarkupBuilder builder, BlockInfo block, FeatureInfo feature, IterationInfo iteration ) {
-        def trCssClassArg = ( feature.skipped ? [ 'class': 'ignored' ] : null )
-        if ( !Utils.isEmptyOrContainsOnlyEmptyStrings( block.texts ) )
-            block.texts.eachWithIndex { blockText, index ->
-                if ( iteration ) {
-                    blockText = stringProcessor.process( blockText, feature.dataVariables, iteration )
+    private void writeBlocks( MarkupBuilder builder, FeatureInfo feature, IterationInfo iteration ) {
+        for ( BlockInfo block in feature.blocks ) {
+            if ( !Utils.isEmptyOrContainsOnlyEmptyStrings( block.texts ) )
+                block.texts.eachWithIndex { blockText, index ->
+                    if ( iteration ) {
+                        blockText = stringProcessor.process( blockText, feature.dataVariables, iteration )
+                    }
+                    writeBlockRow( builder, trCssClass( feature ),
+                            ( index == 0 ? block.kind : 'and' ), blockText )
                 }
-                writeBlockRow( builder, trCssClassArg,
-                        ( index == 0 ? block.kind : 'and' ), blockText )
+            else if ( !hideEmptyBlocks ) {
+                writeBlockRow( builder, trCssClass( feature ), block.kind, '----' )
+            }
+        }
+    }
+
+    private void writeBlocksFromCode( MarkupBuilder builder, FeatureInfo feature, IterationInfo iteration ) {
+        def blocks = codeReader.getBlocks( feature )
+        for ( BlockCode block in blocks ) {
+            def text = iteration && block.text ?
+                    stringProcessor.process( block.text, feature.dataVariables, iteration ) :
+                    ( block.text ?: '' )
+            def blockKind = block.label ?: 'Block:'
+            if ( text ) {
+                writeBlockRowFromCodeAndText( builder, trCssClass( feature ), blockKind, block.statements, text )
+            } else  {
+                writeBlockRowFromCodeOnly( builder, trCssClass( feature ), blockKind, block.statements )
```

Comment:
3 methods: `writeBlockRow`, `writeBlockRowFromCodeAndText` and `writeBlockRowFromCodeOnly`. This is so confusing. Please merge these methods and add arguments if needed to differentiate each case.